### DISCUSSION TITLE
Implement custom JobRunShell

### DISCUSSION THREAD

**Discussion by kvantetore** [quartznet/quartznet#1348] [Q&A] [answered]
In my project, I have a lot of database access (through `SqlClient`), which is, for historical reasons, done through static classes which access a static connection string property. For various reasons I would like to use different connection strings for each quartz job, and thought to implement that through making the connection string `AsyncLocal `and set the connection string before running the job. 

In order for that to work, I believe I need to implement a custom `JobRunShell`, and set the connection string before calling `job.Execute(...)`.

As I brows through the quartz.net source code it seems `JobRunShell `is more or less ready to be customized through `IJobRunShellFactory`. The only piece missing is being able to create supply a custom `IJobRunShellFactory `to the scheduler.

The standard way of supplying custom implementations seems to be by supplying type names to `StdSchedulerFactory` through the `props` argument, and instead of instantiating `StdJobRunShellFactory` directly, use `LoadType(...)` and `InstantiateType(...)` to create the `IJobRunShellFactory`. That does not work well for me, though, as I would like to use the DI mechanisms I already have in place to get the connection string and other settings into my custom `IJobRunShellFactory`.

An alternative way of allowing custom implementations would be to add a `IScheduler.JobRunShellFactory` setter similar to `IScheduler.JobFactory`. 
* `StdScheduler` would call `IJobRunShellFactory.Initialize(...)`  and forward the `JobRunShellFactory` to `QuartzScheduler`, which would update its `QuartzSchedulerResources`. 
* `RemoteScheduler.JobRunShellFactory` would raise an exception similar to `RemoteScheduler.JobFactory`

Does this sound like something you would like to include? Should I submit a PR so we have something concrete to discuss?

**Comment by lahma** [comment, marked as answer]
> In my project, I have a lot of database access (through `SqlClient`), which is, for historical reasons, done through static classes which access a static connection string property. For various reasons I would like to use different connection strings for each quartz job, and thought to implement that through making the connection string `AsyncLocal `and set the connection string before running the job.

So I think I'm still missing the bits how the connection string is determined based on job, is it connection string per job type? What I'm trying to figure if you would be able to utilize job data map or trigger data map to supply the correct value or a key to use for lookup so job would only need to read it in `Execute` using the context's merged job data map. 

Having a `IJobRunShellFactory` actually is from Java land where you could have two implementations being in use, one that uses JTA transactions and one that doesn't. So far .NET version hasn't had a request to allow such configuration, let's see if we have one actual use case here!

**Reply by kvantetore** [reply]
As I have a custom JobFactory using AutoFac to construct my jobs, and the JobFactory also provides a custom logger for each job. I could therefore certainly get the connection string from DI and set it at the beginning of Execute() in each job, but I have a lot of jobs and would rather have the jobs shielded from the messy setup code.

**Reply by kvantetore** [reply]
However, instead of extending the JobRunShell on the "outside", I guess I could extend it on the "inside". I could create my own `class JobRunner : IJob`. Instantiate both `JobRunner`and the actual job class in the JobFactory, and pass the actual job to JobRunner. `JobRunner`'s Execute() could do the necessary setup and then call the actual job.

Actually, I think that might be a cleaner solution than extending JobRunShell. Thank you so much for your help! :)

**Reply by lahma** [reply]
Glad that I was able to be of assistance! I hope you get your approach working smoothly.

**Reply by KoshelevS** [reply]
What are the IJobListener and ITriggerListener for if we still need to have an IJob decorator to do something sophisticated on job start and finish? E.g. if we need to use DI in a ITriggerListener implementation or create an aspect of the system that uses a static AsyncLocal field to store its data.

I see that the current architecture is inherited from Java implementation and it must be hard to modify it. But it is possible to add a few extension points to this architecture? Like an ability to easily plug in my own implementation of IJobRunShellFactory so I can override JobRunShell and plug the extensions I need.

**Reply by lahma** [reply]
I don't oppose such addition and PR. `StdSchedulerFactory` currently has job run shell factory hard-coded, but it could follow the same logic as others allow configuring custom implementation with type, maybe behind key `quartz.jobRunShellFactory.type`. PR's welcome against 3.x branch!
